typed/union.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package typed
+
+import (
+	"fmt"
+	"unicode"
+
+	"sigs.k8s.io/structured-merge-diff/schema"
+	"sigs.k8s.io/structured-merge-diff/value"
+)
+
+var (
+	normalizeUnions = mergeRule(func(w *mergingWalker) {
+		atom, found := w.schema.Resolve(w.typeRef)
+		if !found {
+			panic(fmt.Sprintf("Unable to resolve schema in normalize union: %v/%v", w.schema, w.typeRef))
+		}
+		// Unions can only be in structures
+		if atom.Struct == nil {
+			return
+		}
+		// If the map was removed altogether, let's keep it that way.
+		if w.out == nil {
+			return
+		}
+
+		for _, union := range atom.Struct.Unions {
+			normalizeUnion(union, w.lhs, w.rhs, w.out)
+		}
+	})
+)
+
+func normalizeUnion(union schema.Union, old, new, out *value.Value) {
+	df := ""
+	if union.Discriminator != nil {
+		df = uncapitalize(*union.Discriminator)
+		odv := getDiscriminatorValue(old, df)
+		ndv := getDiscriminatorValue(new, df)
+		if odv != ndv && ndv != "" {
+			for _, field := range union.Fields {
+				if field != ndv {
+					out.MapValue.Delete(field)
+				}
+			}
+			return
+		}
+	}
+
+	ofs := unionFieldsSet(old, union.Fields)
+	nfs := unionFieldsSet(new, union.Fields)
+
+	// Remove fieldset in ofs from nfs
+	for field := range ofs {
+		delete(nfs, field)
+	}
+	// If we have nothing left, then nothing changed,
+	// if we have more than 1, it means that more than 1 field changed
+	if len(nfs) != 1 {
+		return
+	}
+
+	// set discriminator value to only item in nfs
+	dv := ""
+	for v := range nfs {
+		dv = v
+	}
+
+	for field := range ofs {
+		if _, ok := nfs[field]; !ok {
+			out.MapValue.Delete(field)
```

Comment:
Maybe we could use the same
```
            for _, field := range union.Fields {
				if field != ndv {
					out.MapValue.Delete(field)
				}
			}
```
loop from before (or a function that does that) and not worry about ofs or nfs anymore?